db/core/htrie.c

```diff
@@ -86,6 +86,71 @@ tdb_set_bit(unsigned long *bmp, unsigned int nr)
 	set_bit(nr % BITS_PER_LONG, bmp + nr / BITS_PER_LONG);
 }
 
+/**
+ * Reserve the last block of extent for reference table.
+ */
+static void
+tdb_reserve_reftbl(TdbExt *e)
+{
+	set_bit(BITS_PER_LONG - 1, &(e->b_bmp[TDB_BLK_BMP_2L - 1]));
+}
+
+static atomic_t*
+tdb_get_reftbl(TdbHdr *dbh, TdbExt *e)
+{
+	return TDB_PTR(dbh, (TDB_EXT_BASE(dbh, e) + TDB_EXT_SZ) - TDB_BLK_SZ);
+}
+
+/**
+ * Add block to per-CPU freelist.
+ *
+ * NOTE Due to per-CPU nature of freelist, block can be allocated only on CPU
+ * where it was freed. It can cause allocation failures in case when one
+ * CPU frees more blocks then others and free blocks available only in per-CPU
+ * freelists because main memory pool is exhausted.
+ */
+static void
+tdb_freelist_push(TdbHdr *dbh, unsigned long ptr)
+{
+	unsigned long *page_addr = TDB_PTR(dbh, ptr);
+
+	if (!this_cpu_ptr(dbh->pcpu)->freelist) {
+		this_cpu_ptr(dbh->pcpu)->freelist = ptr;
+		*page_addr = 0;
+	} else {
+		*page_addr = this_cpu_ptr(dbh->pcpu)->freelist;
+		this_cpu_ptr(dbh->pcpu)->freelist = ptr;
+	}
+}
+
+static void
+tdb_get_blk(TdbHdr *dbh, unsigned long ptr)
+{
+	TdbExt *e = tdb_ext(dbh, TDB_PTR(dbh, ptr));
+	atomic_t *reftbl = tdb_get_reftbl(dbh, e);
```

Comment:
There is two reasons:
1. Because we want to store reference counter into tdb, that gives an ability to save/restore tdb on/from disk.
2. During system boot, Tempesta reserves memory for TDB. If default allocation failed we rollback to vmalloc. Using refcounter with such pages requires additional logic that we don't have now.